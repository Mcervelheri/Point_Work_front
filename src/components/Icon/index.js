import React, {
    memo, useEffect, useState, useCallback, useMemo,
} from 'react';

import PropTypes from 'prop-types';

export const FontAwesome = 'FontAwesome';
export const Ionicons = 'Ionicons';
export const MaterialIcons = 'MaterialIcons';
export const Octicons = 'Octicons';
export const Typicons = 'Typicons';
export const AntDesign = 'AntDesign';

const getIconFamily = async familyName => {
    let iconFamily;
    switch (familyName) {
        case FontAwesome:
            iconFamily = await import('react-icons/fa');
            break;
        case Typicons:
            iconFamily = await import('react-icons/ti');
            break;
        case Ionicons:
            iconFamily = await import('react-icons/io');
            break;
        case Octicons:
            iconFamily = await import('react-icons/go');
            break;
        case AntDesign:
            iconFamily = await import('react-icons/ai');
            break;
        case MaterialIcons:
        default:
            iconFamily = await import('react-icons/md');
            break;
    }
    return iconFamily;
};

const getIconComponent = (family, name = '') => {
    try {
        const iconName = name.toLowerCase().replace(/[^a-z0-9]/g, '');
        const iconKey = Object.keys(family)
            .find(key => key.substring(2).toLowerCase() === iconName);
        return family[iconKey];
    } catch (ex) {
        console.warn(ex);
        return null;
    }
};

const Icon = memo(({
    family, name, size, style, ...others
}) => {
    const [iconFamily, setIconFamily] = useState(null);

    const requestFamily = useCallback(async () => {
        const iconFam = await getIconFamily(family);
        setIconFamily(iconFam);
    }, [family]);

    useEffect(() => {
        requestFamily();
    }, [requestFamily]);

    const IconComp = useMemo(() => iconFamily && getIconComponent(iconFamily, name), [iconFamily, name]);

    const iconStyles = useMemo(() => ({
        width: size,
        height: size,
        ...style,
    }), [size, style]);

    if (!IconComp) {
        return (
            <span
                style={iconStyles}
                {...others}
            >
                {' '}
            </span>
        );
    }

    return (
        <IconComp
            size={size}
            style={iconStyles}
            {...others}
        />
    );
});

Icon.propTypes = {
    family: PropTypes.oneOf([
        Typicons,
        FontAwesome,
        Ionicons,
        MaterialIcons,
        Octicons,
        AntDesign,
    ]),
    name: PropTypes.string.isRequired,
    size: PropTypes.number,
    style: PropTypes.shape({}),
};

Icon.defaultProps = {
    family: MaterialIcons,
    style: undefined,
    size: undefined,
};

export default Icon;
