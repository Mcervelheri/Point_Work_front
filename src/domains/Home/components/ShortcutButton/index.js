import React, { useCallback } from 'react';

import { useHistory } from 'react-router-dom';

import styles from './styles.module.scss';

const ShortcutButton = ({ title, icon, path }) => {

    const history = useHistory();

    const gotoPath = useCallback(() => {
        history.push(`/app${path}`);
    }, [history, path]);

    return (
        <div>
            <button
                type="button"
                className={styles.button}
                onClick={gotoPath}
            >
                {icon}
            </button>

            <p className={styles.title}>{title}</p>
        </div>
    );
};

export default ShortcutButton;
