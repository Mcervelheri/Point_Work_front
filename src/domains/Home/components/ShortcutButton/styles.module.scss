@import "../../../../theme/colors.scss";
@import "../../../../theme/variables.scss";

.button {
    border: none;
    border-radius: 100%;
    background-color: $color-cyan-7;
    width: 40px;
    height: 40px;
    cursor: pointer;
}

.title {
    margin-top: 13px;
    font-size: $medium-font;
    line-height: 22px;
    text-align: center;
    color: $color-cyan-7;
}
