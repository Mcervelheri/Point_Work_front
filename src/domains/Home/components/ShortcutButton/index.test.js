import React from 'react';

import { createMemoryHistory } from 'history';
import { Router } from 'react-router-dom';
import renderer from 'react-test-renderer';

import Icon, { AntDesign } from '../components/Icon';

import ShortcutButton from './index';

const history = createMemoryHistory({
    initialEntries: [''],
});

test('O Componente <ShortcutButton /> deve ser renderizado na tela corretamente', () => {
    const Element = renderer.create(
        <Router history={history}>
            <ShortcutButton
                path="/path/test"
                icon={(
                    <Icon
                        family={AntDesign}
                        name="FillBank"
                        size={20}
                    />
                )}
                title="titulo test"
            />
        </Router>,

    ).toJSON();
    expect(Element).toMatchSnapshot();
});
