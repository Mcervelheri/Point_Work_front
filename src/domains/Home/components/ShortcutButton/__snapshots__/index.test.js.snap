// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`O Componente <ShortcutButton /> deve ser renderizado na tela corretamente 1`] = `
<div>
  <button
    className="button"
    onClick={[Function]}
    type="button"
  >
    <span
      style={
        Object {
          "height": 20,
          "width": 20,
        }
      }
    >
       
    </span>
  </button>
  <p
    className="title"
  >
    titulo test
  </p>
</div>
`;
